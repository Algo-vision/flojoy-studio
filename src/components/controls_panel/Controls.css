.App-controls-panel {
    height: 100vh; 
}

.App-controls-header {
  display: flex;
  flex-wrap: wrap;
  flex-direction: row;
  justify-content: start;
  align-items: auto;
  align-content: start; 
}

.ctrl-output {
  flex: 0 0 auto;
}

.ctrl-output, .ctrl-input {
  padding: 20px;
  border: 0.5px solid #ccc;
  border-radius: 3px;
  position: relative;
  resize: both;
  cursor: pointer;
  -webkit-user-drag: element;
  margin-top: 10px;
}

.ctrl-output {
  margin: 20px 0 0 20px;
  min-height: 300px;
  height: 25%;
  width: 45%;
}

.ctrl-input {
  min-height: 100px;
  height: 10%;
  width: 100%;
}

.ctrl-close-btn {
  margin: 5px;
  cursor: pointer;
  position: absolute;
  top: 0;
  right: 0;
  border: 0;
  background: transparent;
}

.ctrl-meta {
  font-size: 10px;
  text-align: left;
  position: absolute;
  bottom: 0
}

.input-header, .output-header {
  padding: 10px;
  display: inline-block;
}

.ctrl-inputs-sidebar, .input-header {
  min-width: 200px;
  max-width: 400px;
  width: 25%;  border-right: 1px solid #ccc;
  display: inline-block;
}

.ctrl-inputs-sidebar {
  height: 100%;
  padding: 10px;
}

.ctrl-outputs-canvas {
  display: flex;
  background-size: .75em .75em;  
  background-position: 0 -0.5em;
  flex-wrap: wrap;
  flex-direction: row;
  justify-content: start;
  align-items: auto;
  align-content: start;
  min-height: 800px;
}

.ctrl-outputs-container, .output-header {
  display: inline-block;
  width: 75%;
}
.ctrl-outputs-container {
  vertical-align: top;
  height: 100%;
}