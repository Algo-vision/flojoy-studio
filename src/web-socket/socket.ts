export class WebSocketServer {
  private server: WebSocket;
  private pingResponse: any;
  private heartbeatResponse: any;
  constructor(url: string, pingResponse: any, heartbeatResponse: any) {
    this.pingResponse = pingResponse;
    this.heartbeatResponse = heartbeatResponse;
    this.server = new WebSocket(url);
    this.init();
  }
  init() {
    console.log(" socket readystate: ", this.server.readyState);
    this.server.onmessage = (ev) => {
      let data = JSON.parse(ev.data);
      // console.log("data received: ", data.type === "heartbeat_response");
      switch (data.type) {
        case "heartbeat_response":
          if (this.heartbeatResponse) {
            if (this.pingResponse) {
              this.pingResponse(data.msg);
            }
            this.heartbeatResponse(data);
          }
          break;
        case "ping_response":
          if (this.pingResponse) {
            this.pingResponse(data.msg);
          }
          break;
        case "connection_established":
          this.server.send(
            JSON.stringify({
              type: "ping",
            })
          );
          break;

        default:
          console.log(" default data type: ", data);
          break;
      }
    };
  }
  disconnect() {
    this.server.close(0);
  }
  emit(data: string) {
    this.server.send(
      JSON.stringify({
        type: data,
      })
    );
  }
  isConnected() {
    return this.server.readyState;
  }
}
