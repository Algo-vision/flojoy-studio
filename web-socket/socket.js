const {Server} = require('socket.io');

class WebSocketServer {
    #server;
    #socket;
    pingResponse
    constructor(server, pingResponse){
        this.pingResponse = pingResponse;
        this.#server = new Server(server,{cors:{
            origin: '*'
        }});
        this.init();
    }
    init(){
        this.#server.on('connection', (socket)=> {
            this.#socket = socket;
            this.#socket.on('ping', ()=>{
                if(this.pingResponse){
                   this.pingResponse(this.#socket)
                }
            })
        })
    }
    sendSystemStatus(status){
        if(this.#socket){
            this.#socket.emit('system-status', status)
        }
    }
    returnResult(result){
        if(this.#socket){
            this.#socket.emit('program-results', result)
        }
    }
}
module.exports = WebSocketServer;