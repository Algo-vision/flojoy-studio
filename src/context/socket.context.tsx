import { createContext, useEffect, useRef, useState } from "react";
import { WebSocketServer } from "../web-socket/socket";

export const SocketContext = createContext<any>(null);

type States = {
  programResults: any;
  runningNode: string;
  serverStatus: string;
  failedNodes: any[];
  failureReason: any[];
};
export const SocketContextProvider = ({ children }) => {
  const socket = useRef<WebSocketServer>();
  const [states, setStates] = useState<States>({
    programResults: {},
    runningNode: "",
    serverStatus: "Connecting to server...",
    failedNodes: [],
    failureReason: [],
  });
  const handleStateChange = (state: keyof States) => (value: any) => {
    setStates((prev) => ({
      ...prev,
      [state]: value,
    }));
  };
  useEffect(() => {
    if (!socket.current) {
      socket.current = new WebSocketServer({
        url: "ws://localhost:8000/ws/socket-server/",
        pingResponse: handleStateChange('serverStatus'),
        heartbeatResponse: handleStateChange('programResults'),
        runningNode: handleStateChange('runningNode'),
        failedNodes: handleStateChange('failedNodes'),
        failureReason: handleStateChange('failureReason')
      });
    }
  }, []);
  return (
    <SocketContext.Provider
      value={{ states }}
    >
      {children}
    </SocketContext.Provider>
  );
};
