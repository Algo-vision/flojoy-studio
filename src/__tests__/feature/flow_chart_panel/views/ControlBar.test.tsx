import { render, screen } from "@testing-library/react";
import Controls from "@src/feature/flow_chart_panel/views/ControlBar";
import { IServerStatus } from "@src/context/socket.context";

// Mock uuid
jest.mock("uuid", () => ({
  v4: () => "randomId",
}));

// Mock useFlowChartState hook
jest.mock("@src/hooks/useFlowChartState", () => ({
  useFlowChartState: () => ({
    isEditMode: false,
    setIsEditMode: jest.fn(),
    rfInstance: null,
    openFileSelector: jest.fn(),
    saveFile: jest.fn(),
    nodes: [],
    setNodes: jest.fn(),
  }),
}));
// Mock useSocket hook
jest.mock("@src/hooks/useSocket", () => ({
  useSocket: () => ({
    states: {
      socketId: "socketId",
      setProgramResults: jest.fn(),
      serverStatus: IServerStatus.CONNECTING,
    },
  }),
}));

// Mock saveFlowChartToLocalStorage, saveAndRunFlowChartInServer, and cancelFlowChartRun functions
jest.mock("@src/services/FlowChartServices", () => ({
  saveFlowChartToLocalStorage: jest.fn(),
  saveAndRunFlowChartInServer: jest.fn(),
  cancelFlowChartRun: jest.fn(),
}));

// Mock AddNodeModal component
jest.mock("@src/feature/flow_chart_panel/views/AddNodeModal", () => ({
  __esModule: true,
  default: jest.fn(() => <div>AddNodeModal</div>),
}));

// Mock KeyboardShortcutModal component
jest.mock("@src/feature/flow_chart_panel/views/KeyboardShortcutModal", () => ({
  __esModule: true,
  default: jest.fn(() => <div>KeyboardShortcutModal</div>),
}));
// Mock DropDown component
jest.mock("@src/feature/common/dropdown/DropDown", () => ({
  __esModule: true,
  default: jest.fn(() => <div>Dropdown</div>),
}));

describe("Controls", () => {
  it("should render correctly", () => {
    const { container } = render(
      <Controls theme="dark" activeTab="visual" setOpenCtrlModal={jest.fn()} />
    );

    expect(screen.getByTestId("btn-play")).toBeInTheDocument();
    expect(screen.getByTestId("btn-play")).toBeDisabled();
    expect(screen.queryByTitle("Cancel Run")).not.toBeInTheDocument();
    expect(screen.getByTestId("add-node")).toBeInTheDocument();
    expect(screen.queryByTestId("add-ctrl")).not.toBeInTheDocument();
    expect(screen.getByText("Dropdown")).toBeInTheDocument();
    expect(screen.getByText("AddNodeModal")).toBeInTheDocument();
    expect(container).toMatchSnapshot("__main__");
  });
  it("should show the Add node button when active tab is 'visual'", () => {
    render(
      <Controls theme="dark" activeTab="visual" setOpenCtrlModal={jest.fn()} />
    );
    expect(screen.getByTestId("add-node")).toBeInTheDocument();
    expect(screen.queryByTestId("add-ctrl")).not.toBeInTheDocument();
  });

  it("should show the Edit button when active tab is 'panel'", () => {
    render(
      <Controls theme="dark" activeTab="panel" setOpenCtrlModal={jest.fn()} />
    );
    expect(screen.queryByTestId("add-node")).not.toBeInTheDocument();
    expect(screen.getByTestId("operation-switch")).toBeInTheDocument();
  });
  it("should not show the Add node or add ctrl button when active tab is 'debug'", () => {
    render(
      <Controls theme="dark" activeTab="debug" setOpenCtrlModal={jest.fn()} />
    );
    expect(screen.queryByTestId("add-node")).not.toBeInTheDocument();
    expect(screen.queryByTestId("add-ctrl")).not.toBeInTheDocument();
  });
});
