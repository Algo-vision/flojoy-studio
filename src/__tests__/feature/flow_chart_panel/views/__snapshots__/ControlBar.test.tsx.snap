// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls should renderWithTheme correctly: __main__ 1`] = `
<div>
  <div
    class="mantine-1mo5lls"
  >
    <button
      class="mantine-8mgm8e"
      data-cy="btn-play"
      data-testid="btn-play"
      disabled=""
      title="Server is offline"
    >
      <svg
        fill="none"
        height="11"
        viewBox="0 0 9 11"
        width="9"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8.5 4.63397C9.16667 5.01887 9.16667 5.98113 8.5 6.36603L1.75 10.2631C1.08333 10.648 0.25 10.1669 0.25 9.39711L0.25 1.60289C0.25 0.833085 1.08333 0.35196 1.75 0.73686L8.5 4.63397Z"
          fill="#99f5ff"
        />
      </svg>
      <span>
        Play
      </span>
    </button>
    <div>
      Dropdown
    </div>
    <div>
      KeyboardShortcutModal
    </div>
  </div>
  ,
</div>
`;

exports[`Controls should show the Edit button when active tab is 'panel': __switch_btn__ 1`] = `
<div>
  <div
    class="mantine-1mo5lls"
  >
    <button
      class="mantine-8mgm8e"
      data-cy="btn-play"
      data-testid="btn-play"
      disabled=""
      title="Server is offline"
    >
      <svg
        fill="none"
        height="11"
        viewBox="0 0 9 11"
        width="9"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8.5 4.63397C9.16667 5.01887 9.16667 5.98113 8.5 6.36603L1.75 10.2631C1.08333 10.648 0.25 10.1669 0.25 9.39711L0.25 1.60289C0.25 0.833085 1.08333 0.35196 1.75 0.73686L8.5 4.63397Z"
          fill="#99f5ff"
        />
      </svg>
      <span>
        Play
      </span>
    </button>
    <div>
      Dropdown
    </div>
    <div
      class="mantine-15fjqzh"
    >
      <div
        class="mantine-Text-root mantine-1gamrgd"
        data-cy="operation-switch"
        data-testid="operation-switch"
      >
        Edit
      </div>
      <div
        data-testid="react-switch"
      />
    </div>
    <div>
      KeyboardShortcutModal
    </div>
  </div>
  ,
</div>
`;
